### DISCUSSION TITLE
Advice?

### DISCUSSION THREAD

**Discussion by uudruid74** [CIDRAM/CIDRAM#587] [Q&A] [answered]
So, I have CIDRAM in front of Opencart.  I am getting idiots signing up for accounts with random names, then they make a crap ton of  inquiry emails with just random trash as they attempt to do whatever, then they try returning products they never purchased (why the cart allows people with no order history to even click the return link I will try and fix today by hacking the code).  I'll get 20 emails in a row!

Anyway, my customer list is being filled up with trash accounts because the cart is too stupid to verify the email address (opencart is a joke, huge mistake, but too late now).

Anyone have a way to block all non-US IPs?  Can I add a rule that would block people that click the contact form submit button more than twice?  Or would the rule penalize them at the first click?

Also, what is the best way to make CIDRAM a lot more aggressive in how quickly it blocks?  Recommendations on how to best move forward with this would be appreciated.  Thanks.

**Comment by uudruid74** [comment]
Is there a way to block people that hit the same link twice within 1 minute?  Every one of these idiots hits the contact page twice in 1 minute with the same data.

**Comment by Maikuolan** [comment, marked as answer]
> Anyone have a way to block all non-US IPs?

It's certainly possible to have CIDRAM block all non-US IPs, or all non-anywhere-else IPs, for that matter, but with a few caveats:

To know the country of origin of an inbound request, CIDRAM needs to be able to look up the IP address for that request at an external service via one of its modules (e.g., the IP-API or BGPView modules). When one of those modules are installed and active, the lookup strategy used (i.e., when lookups would be performed, and under what circumstances) can be configured via the configuration page for whether to perform lookups for every request or just for sensitive pages only (e.g., login pages, registration forms, etc), and by default, is set to perform lookups for sensitive pages only.

CIDRAM doesn't store comprehensive, internet-wide information of that nature internally, partly due to the size such a data set would reach and the footprint such a size would have on CIDRAM's overall size, but mostly because a fair amount of the data in such a data set would stale on a nearly daily basis, and maintaining that kind of data to try to keep it all up to date all the time is a kind of nightmare I'd rather not deal with myself. Thankfully, there are enough external services providing their own APIs around that I don't need to.

For websites with heavy traffic, I wouldn't generally recommend performing such lookups or other external requests at *every* request to said websites due to the likelihood of a strongly noticeable negative performance impact occurring as a result of that, plus the potential for those external services to rate limit or block said websites as a result (performing lookups at every request, for a website with heavy traffic, is likely to see daily lookup quota being eaten up super quickly, or could, in theory, even potentially be seen by them as a kind of DDoS attack if enough such heavy-traffic websites were to do so, though to my knowledge, that hasn't yet ever happened before). It might be okay for smaller websites with particularly low levels of traffic though.

The downside to only performing said lookups at sensitive page requests, of course, is that any rules based on the information derived from those lookups won't come into effect until after such a lookup has occurred (so potentially seeing such IPs never being blocked until after they've clicked on a page where the lookup would normally be triggered).

Any such service also isn't always 100% accurate in the information that it returns, which *might* occasionally result in false positives or missed detections. With that in mind, when acting on that information, it might be worth enabling some of CIDRAM's reCAPTCHA/hCAPTCHA functionality, in order to hopefully allow any people blocked by mistake to regain access to the website again (though, a person from one country is rarely any different than a person from any other country when it comes to one's ability to solve a CAPTCHA, so that may need to be judged on a case-by-case basis, depending on where one's stance is in that regard).

That all said, if you're okay with performing lookups to at least one of those external services, and with the possibility of occasional false positives or missed detections, from there, the easiest way forward would be by creating a new auxiliary rule via the auxiliary rules page to block any requests whose country code lookups don't match the US, like so:

![CCUS](https://github.com/CIDRAM/CIDRAM/assets/12571108/ae376d6f-f960-4eac-bb49-a623cf73f2c7)

**Comment by Maikuolan** [comment]
> Can I add a rule that would block people that click the contact form submit button more than twice? Or would the rule penalize them at the first click?

To know whether someone has clicked on something twice, or whether a particular page has been accessed twice, CIDRAM would likely need to be counting that information, to know that something has happened two times, and tracking how many times each user, or each encountered IP address, has accessed a given page (currently, CIDRAM doesn't do this, and the only information it tracks from specific, individual IP addresses, without including any information contained by CIDRAM's block event logs and such, is how many past infractions an address has occurred, when to expire that information, or, when rate limiting is enabled via the configuration, and depending on how it's configured, how many requests have been made to a website more broadly by an address, or how much total bandwidth more broadly that address has consumed at a website).

Penalising them at the first click is doable and easy enough (and I assume, not what you want here), but penalising them at only the *second* click while *not* doing so at the first is currently difficult due to CIDRAM not tracking how many times any particular specific page is accessed or particular specific event triggered from any given IP address.

I'll need to think on that a bit to try to figure out a solution there.

Anyway, it's just a little passed 1AM now at my local time and I'm starting to get a bit too foggy for reliably coherent replies, so I'll leave addressing the other points here for after I've had some sleep (or someone else can chime in if they feel like doing so). :-)

**Comment by uudruid74** [comment]
I get the million lookups, and can look at more efficient blocking later.  I get very little traffic (a problem I'm working on!) at the moment.  The sledgehammer approach of US only is working wonderfully.  Wish I had something like this 20 years ago!   Thank you so much!

**Comment by Maikuolan** [comment]
Happy to help. :-)
